mesos-client/src/main/scala/mesosphere/mesos/client/MesosClient.scala

```diff
@@ -354,200 +135,276 @@ trait MesosApi {
+
+              (response, ConnectionInfo(url, streamId.value()))
+            case StatusCodes.TemporaryRedirect =>
+              val leader = new URI(response.header[headers.Location].get.value())
+              logger.warn(s"New mesos leader available at $leader")
+              // Update the context with the new leader's host and port and throw an exception that is handled in the
+              // next `recoverWith` stage.
+              response.discardEntityBytes()
+              throw new MesosRedirectException(leader)
+            case _ =>
+              response.discardEntityBytes()
+              throw new IllegalArgumentException(s"Mesos server error: $response")
+          }
+        }
+        .recoverWithRetries(conf.redirectRetries, {
+          case MesosRedirectException(leader) => mesosHttpConnection(leader, redirectRetries)
+        })
+
+    val initialUrl = new java.net.URI(s"http://${conf.master}")
+
+    val httpConnection: Source[(HttpResponse, ConnectionInfo), NotUsed] =
+      mesosHttpConnection(initialUrl, conf.redirectRetries)
+
+    val eventReader = Flow[HttpResponse]
+      .flatMapConcat(_.entity.dataBytes)
+      .via(RecordIOFraming.scanner())
+      .via(eventDeserializer)
+      .via(log("Received mesos Event: "))
+      .idleTimeout(conf.idleTimeout)
+      .buffer(conf.sourceBufferSize, OverflowStrategy.backpressure)
+
+    val eventsOutputSink = Sink.asPublisher[Event](false)
+
+    val sharedKillSwitch = KillSwitches.shared(s"MesosClient-${conf.master}")
+
+    val graph = GraphDSL.create(
+      httpConnection, subscribedWatcher, Sink.head[ConnectionInfo], sharedKillSwitch.flow[Event], eventsOutputSink)(
+      { (_, subscribedF, connectionInfoF, _, eventsOutputPublisher) =>
+        for {
+          subscribed <- subscribedF
+          connectionInfo <- connectionInfoF,
+        } yield {
+          new MesosClientImpl(sharedKillSwitch, subscribed, connectionInfo, Source.fromPublisher(eventsOutputPublisher))
+        }
+      }
+    ) { implicit b =>
+        { (httpConnectionShape, subscribedWatchedShape, connectionInfoWatcher, killSwitch, eventsOutputSinkShape) =>
+          import GraphDSL.Implicits._
+
+          val unzip = b.add(Unzip[HttpResponse, ConnectionInfo])
+          val eventBroadcast = b.add(Broadcast[Event](2, eagerCancel = false))
+
+          // Wire output events
+          httpConnectionShape ~> unzip.in
+          unzip.out0.via(eventReader) ~> killSwitch ~> eventBroadcast
+          unzip.out1 ~> connectionInfoWatcher
+
+          eventBroadcast.out(0).take(1) ~> subscribedWatchedShape
+          /* we use detach to prevent a deadlock situation. "drop(1)" does not initiate a pull, so we use detach to
+           * preemptively pull an element so that the first broadcast output can receive an element. */
+          eventBroadcast.out(1).drop(1).detach ~> eventsOutputSinkShape
+          ClosedShape
+        }
+      }
```

Comment:
Unfortunately I don't have enough knowledge to judge this graph. However, why do we need `subscribedWatcher`? Would it not be enough to take the first element? Rust streams have `into_future` which give `Future[(FirstItem, Stream)]`.

Reply:
Yeah... this can be improved. I didn't think of it at the time but there is prefixAndTail which will achieve something similar. I'll rework it.

Reply:
`prefixAndTail` is exactly what I meant 🙂.

Reply:
Glad you said something; prefixAndTail is substantially simpler.